src/arg_parsers/filename.rs

```diff
@@ -20,78 +20,123 @@ use crate::{arg_parse_err::ArgParseErr, error::MagickError, wm_err};
 
 use super::{Geometry, ResizeGeometry};
 
+#[derive(Clone, Debug, Default, PartialEq)]
+pub enum Location {
+    Path(PathBuf),
+    #[default]
+    Stdio,
+}
+
+impl Location {
+    pub fn from_arg(arg: &OsStr) -> Self {
+        if matches!(arg.as_encoded_bytes(), b"" | b"-") {
+            Self::Stdio
+        } else {
+            Self::Path(PathBuf::from(arg))
+        }
+    }
+}
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct InputFileArg {
-    pub path: PathBuf,
+    pub location: Location,
     pub format: Option<ImageFormat>,
     pub read_mod: Option<ReadModifier>,
 }
 
 impl InputFileArg {
     pub fn parse(input: &OsStr) -> Result<Self, MagickError> {
```

Comment:
I rewrote most this method because adding Location handling to the existing code made it really hard to follow. I think this matches ImageMagick's behaviour but more testing would be great. I also added unit tests for some of the edge cases.